mono/utils/mono-os-wait-win32.c

```diff
@@ -0,0 +1,210 @@
+#include <mono/utils/mono-os-wait.h>
+#include <mono/utils/mono-threads.h>
+#include <mono/utils/mono-threads-debug.h>
+
+#define THREAD_WAIT_INFO_CLEARED 0
+#define THREAD_WAIT_INFO_ALERTABLE_WAIT_SLOT 1
+#define THREAD_WAIT_INFO_PENDING_INTERRUPT_APC_SLOT 2
+#define THREAD_WAIT_INFO_PENDING_ABORT_APC_SLOT 4
+
+static inline void
+reqeust_interrupt (gpointer thread_info, HANDLE native_thread_handle, gint32 pending_apc_slot, PAPCFUNC apc_callback, DWORD tid)
+{
+	/*
+	* On Windows platforms the async interrupt/abort requests might need to process
+	* APC on target thread before the thread can return back from OS wait calls
+	* and complete the mono interrupt/abort request. In such cases, just keep on
+	* queuing APC over again again will flood the APC queue preventing the target thread
+	* to return from it's alertable OS wait call. This check makes sure not to issue an APC
+	* as long as there is a pending requests in requested APC slot for the targeted thread.
+	* NOTE, this code will executed regardless if thread is currently in an alertable wait or not.
+	* This is done to prevent races between interrupt/abort occurring just before thread enters an
+	* alertable wait. Threads entering waits already need to handle WAIT_IO_COMPLETION scenarios and
+	* if that happens due to a previous pending interrupt/abort APC, the calling thread should already
+	* have logic to restart the alertable wait operation.
+	*/
+	MonoThreadInfo *info = (MonoThreadInfo *)thread_info;
+	gboolean queue_apc = FALSE;
+
+	while (!queue_apc) {
```

Comment:
The usual way in the runtime of formatting a `InterlockedCompareExchange` loop is the following:

```
gint old_wait_info, new_wait_info;
do {
    old_wait_info = info->thread_wait_info;
    if (old_wait_info & pending_apc_slot)
        return;

    new_wait_info = old_wait_info | pending_apc_slot;
} while (InterlockedCompareExchange (&info->thread_wait_info, new_wait_info, old_wait_info) != old_wait_info);

THREADS_INTERRUPT_DEBUG ("%06d - Interrupting/Aborting syscall in thread %06d", GetCurrentThreadId (), tid);
QueueUserAPC (apc_callback, native_thread_handle, (ULONG_PTR)NULL);
```